File: leetcode/721.accounts-merge/721.accounts-merge.py
#
# @lc app=leetcode id=721 lang=python3
#
# [721] Accounts Merge
#

# @lc code=start
class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        from collections import defaultdict

        visited = set()
        d = defaultdict(list)

        def dfs(email, mergedAcc):
            if email in visited:
                return
            visited.add(email)
            mergedAcc.append(email)

            for nb in d[email]:
                dfs(nb, mergedAcc)

        for account in accounts:
            firstEmail = account[1]

            for i in range(1, len(account)):
                d[firstEmail].append(account[i])
                d[account[i]].append(firstEmail)

        res = []
        for account in accounts:
            name = account[0]
            firstEmail = account[1]

            if firstEmail in visited:
                continue

            mergedAcc = []
            dfs(firstEmail, mergedAcc)
            mergedAcc = [name] + sorted(mergedAcc)
            res.append(mergedAcc)

        return res


# @lc code=end

File: leetcode/721.accounts-merge/721.accounts-merge.cpp
/*
 * @lc app=leetcode id=721 lang=cpp
 *
 * [721] Accounts Merge
 */

// @lc code=start
class Solution
{
public:
    vector<vector<string>> accountsMerge(vector<vector<string>> &accounts)
    {
        class Solution
        {
        public:
            unordered_set<string> visited;
            unordered_map<string, vector<string>> al;

            void dfs(vector<string> &mergedAccount, string email)
            {
                if (visited.find(email) != visited.end())
                    return;
                visited.insert(email);
                mergedAccount.push_back(email);

                for (auto nb : al[email])
                {
                    dfs(mergedAccount, nb);
                }
            }

            vector<vector<string>> accountsMerge(vector<vector<string>> &accounts)
            {

                for (auto account : accounts)
                {
                    string first = account[1];

                    for (int i = 2; i < account.size(); ++i)
                    {
                        string email = account[i];
                        al[first].push_back(email);
                        al[email].push_back(first);
                    }
                }

                vector<vector<string>> res;
                for (auto account : accounts)
                {
                    string name = account[0];
                    string first = account[1];

                    if (visited.find(first) != visited.end())
                        continue;

                    vector<string> mergedAccount;
                    mergedAccount.push_back(name);
                    dfs(mergedAccount, first);
                    sort(mergedAccount.begin() + 1, mergedAccount.end());
                    res.push_back(mergedAccount);
                }

                return res;
            }
        };
    }
};
// @lc code=end
